libp2p/protocols/kademlia/find.nim

```diff
@@ -7,115 +7,140 @@
 # This file may not be copied, modified, or distributed except according to
 # those terms.
 
-import std/[tables, sequtils, sets]
+import std/[tables, sequtils, sets, algorithm]
 import chronos, chronicles, results
-import ../../[peerid, switch, multihash]
+import ../../[peerid, peerinfo, switch, multihash]
 import ../protocol
-import ./[routingtable, lookupstate, protobuf, types]
+import ./[routingtable, protobuf, types]
 
-proc checkConvergence(state: LookupState, me: PeerId): bool {.raises: [], gcsafe.} =
-  let ready = state.activeQueries == 0
-  let noNew = state.selectAlphaPeers().filterIt(me != it).len == 0
-  return ready and noNew
+type LookupState* = object
+  kad: KadDHT
+  targetId: Key
+  shortlist: Table[PeerId, XorDistance]
+  queried: HashSet[PeerId]
 
-proc waitRepliesOrTimeouts(
-    pendingFutures: Table[PeerId, Future[Message]]
-): Future[(seq[Message], seq[PeerId])] {.async: (raises: [CancelledError]).} =
-  await allFutures(pendingFutures.values.toSeq())
+proc updateShortlist*(
+    state: var LookupState, msg: Message
+): seq[PeerInfo] {.raises: [].} =
+  var newPeerInfos: seq[PeerInfo]
 
-  var receivedReplies: seq[Message] = @[]
-  var failedPeers: seq[PeerId] = @[]
-
-  for (peerId, replyFut) in pendingFutures.pairs:
-    try:
-      receivedReplies.add(await replyFut)
-    except CatchableError:
-      failedPeers.add(peerId)
-      error "Could not send find_node to peer", peerId, err = getCurrentExceptionMsg()
-
-  return (receivedReplies, failedPeers)
-
-proc sendFindNode*(
-    kad: KadDHT, peerId: PeerId, addrs: seq[MultiAddress], targetId: Key
+  for newPeer in msg.closerPeers:
+    let pid = PeerId.init(newPeer.id).valueOr:
+      continue
+    if not state.shortlist.contains(pid):
+      state.shortlist[pid] =
+        xorDistance(pid, state.targetId, state.kad.rtable.config.hasher)
+
+      let peerInfo = PeerInfo(peerId: pid, addrs: newPeer.addrs)
+      newPeerInfos.add(peerInfo)
+
+  return newPeerInfos
+
+proc selectAlphaPeers*(state: LookupState): seq[PeerId] =
+  var peers: seq[PeerId] = @[]
+
+  # sort by closer distance first
+  let sortedPids = state.shortlist
+    .pairs()
+    .toSeq()
+    # sort by distance (closer first)
+    .sortedByIt(it[1])
+    # get pid
+    .mapIt(it[0])
+    # filter out already queried pids
+    .filterIt(not state.queried.contains(it))
+    # filter out our own pid
+    .filterIt(it != state.kad.switch.peerInfo.peerId)
+
+  # get alpha peers
+  for pid in sortedPids:
+    peers.add(pid)
+    if peers.len() >= state.kad.config.alpha:
+      break
+  peers
+
+proc init*(T: type LookupState, kad: KadDHT, targetId: Key): T =
```

Comment:
`targetId` is of type `Key`....

shouldn't it then be `targetKey` or  `targetId: Id`?